### PR TITLE
dev/financial#20 - Extend pay now functionality for partially paid contribut…

### PR DESCRIPTION
**Pull request by jitendrapurohit**
…ions

Overview
----------------------------------------
Online paynow functionality for partially paid contributions.

Before
----------------------------------------
Partial Paid contribution can only be completed from backend pages.

After
----------------------------------------
Partial Paid has a pay now link on user dashboard which navigates to a payment page for online completion.

image

Payment page list all the price field and the balance due amount.

image

Comments
----------------------------------------
Added unit test.

----------------------------------------

Gitlab issue https://lab.civicrm.org/dev/financial/issues/20

### CODE REVIEW

**Comment by jitendrapurohit** [comment]
ping @monishdeb @eileenmcnaughton

**Comment by monishdeb** [comment]
@jitendrapurohit there is a style warning, can you please fix that?

**Comment by jitendrapurohit** [comment]
@monishdeb Done 👍

**Comment by magnolia61** [comment]
I will test now and feedback in a bit

**Comment by magnolia61** [comment]
Very nice and works mostly as expected, I have just a few suggestions.
1. For a clear user interface and maintainability + because this block became recently available I suggest the use of the Payment Summary block which is also in the contribution view (or otherwise add the 'amount paid'. 
pp_contribution_paynow_block
2. The Contribution information shows a full history of all changes, I believe it should only show items that add to the total amount (would that be the same as QTY > 0?)
pp_all-lineitems_also_notselected
3. When a users clicks the PayNow link after it has already been paid a confusing screen is shown. But an error message is shown AND the price set of the Contribution page that was configured.
pp_paynow_error_alreadypaid
I think the previous behaviour (transparent redirect to the homepage) also is confusing. Instead a helpful message should be shown to the end-user. A related solution to this problem has been submitted as a PR in https://lab.civicrm.org/feature-requests/user-interface/issues/1 which dealt with expired checksum links.

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit related test fails?

**Comment by magnolia61** [comment]
BTW. of my earlier 3 remarks 1 and 2 are easy I think but also 'nice to have' so no problem if the are just documented here. Only #3 is really in need of improvement as both transparent redirect to the homepage as the mix of a message and a price set are both confusing for end users.

**Comment by jitendrapurohit** [comment]
Thanks for the review and UI improvement suggestion @magnolia61 👍 

Re 2/ we don't list the price field on the main page if the user has not opted for it. You screenshot seems to be a result of `Change Selection` made by the user so that the one which is removed is shown as a disabled row in the price field section. I think it adds more info on the page to let user know of the selections updated by them. Same is done on the view screen at the backend. 

Can anyone @eileenmcnaughton @colemanw do a +1 to @magnolia61 suggestion so that I can proceed in implementing 1 and 2 in a separate ticket?

For 3 - PayNow link is mostly clicked by the user from user dashboard page. If an anonymous visits the paynow link - the error is simply not shown to the user as civicrm doesn't display errors on non-civi pages. Hence, a most common page is chosen here(default invoice payment page URL without ccid or cid) for the redirection so that the user know what was wrong with their page load.

File: CRM/Core/BAO/FinancialTrxn.php [hunk 1]
```diff
@@ -511,14 +511,16 @@ public static function getPartialPaymentWithType($entityId, $entityName = 'parti
    */
   public static function getTotalPayments($contributionId) {
     $statusId = CRM_Core_PseudoConstant::getKey('CRM_Contribute_BAO_Contribution', 'contribution_status_id', 'Completed');
+    $refundStatusId = CRM_Core_PseudoConstant::getKey('CRM_Contribute_BAO_Contribution', 'contribution_status_id', 'Refunded');
 
     $sql = "SELECT SUM(ft.total_amount) FROM civicrm_financial_trxn ft
       INNER JOIN civicrm_entity_financial_trxn eft ON (eft.financial_trxn_id = ft.id AND eft.entity_table = 'civicrm_contribution')
-      WHERE eft.entity_id = %1 AND ft.is_payment = 1 AND ft.status_id = %2";
+      WHERE eft.entity_id = %1 AND ft.is_payment = 1 AND ft.status_id IN (%2, %3)";
```
Comment (jitendrapurohit) [inline CRM/Core/BAO/FinancialTrxn.php]: @monishdeb @eileenmcnaughton @pradpnayak seems calculating total payment from `financial_trxn` table should also consider the negative amount value with refund status? We do the same in [`getPartialPaymentWithType`](https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/BAO/FinancialTrxn.php#L481) function. Thoughts - if this seems fine to you?

**Comment by magnolia61** [comment]
@jitendrapurohit About #3: We mostly use the constructed Pay Now link that we send by email of have available in a view(block) for them. So for us there are not really anonymous users, because people click the link with the checksum.I'll try to describe what happens in real-life and do some suggestions in order of preferability:

A users clicks Pay now link from email: sees price set and message (confusing, maybe the price set does not have anything to do with the payment (there is only 1 default contribution page but many payment scenarios, different events, memberships), 
Result: the user does not know what to do in this case (see screenshot above at #3). Especially the price set that is shown is confusion because the users does not know what to do with that. Should he/she select something, and what happens when he/she does?

Suggestion #1: Best would be only the message (no confusing price set that should be ignored)
Suggestion #2: Otherwise: redirect to homepage without a message (also bad but less confusing)

**Comment by magnolia61** [comment]
> Suggestion #1: Best would be only the message (no confusing price set that should be ignored)
> Suggestion #2: Otherwise: redirect to homepage without a message (also bad but less confusing)

Would it be possible to implement one of my two suggestions to proceed. The solution to show the default contrib page is really too confusing for end users. If my 1st suggestion would require too much work please just revert it to redirect to the homepage in order for this PR to be able to get merged :-)

**Comment by jitendrapurohit** [comment]
@magnolia61 I've updated the PR to display an error message instead of redirecting to any page. Can you pls give it a final testing?

Thanks :-)

**Comment by magnolia61** [comment]
@jitendrapurohit it now shows an error message when logged in, and transparently redirects to the homepage (without  a message) when a Pay Now checksum is used.

While this solution lacks communication to the checksumlink users, I believe for the time being this is no reason to uphold this PR. Great work on a alert message solution for anonymous users has successfully been done at https://github.com/civicrm/civicrm-core/pull/12724 (yet to be merged)

What is disturbing though is the fatal error when using the Pay Now link from the user dashboard (accessed by a checksum link)
screenshot from 2018-08-24 15-08-49

I opened a separate issue for that at https://lab.civicrm.org/dev/core/issues/346 but I think it is closely related to this PR

File: CRM/Contribute/Form/Contribution/Main.php [hunk 2]
```diff
@@ -1310,20 +1310,30 @@ public function assignFormVariablesByContributionID() {
       return;
     }
     if (!$this->getContactID()) {
-      CRM_Core_Error::statusBounce(ts("Returning since there is no contact attached to this contribution id."));
+      CRM_Core_Error::fatal(ts("Returning since there is no contact attached to this contribution id."));
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/Contribution/Main.php]: We shouldn't be adding in fatals. It's a deprecated function.

Comment (jitendrapurohit) [reply CRM/Contribute/Form/Contribution/Main.php]: Modified it to use an exception instead of the above 👍

**Review by magnolia61 — CHANGES_REQUESTED** [review]
To include balance on the user dashboard contribution block I added the following in 
templates/CRM/Contribute/Page/UserDashboard.tpl
ln34:` <th>{ts}Balance Due{/ts}</th>`
ln56: `<td>{$row.balance_due}</td>`
Can you agree and would you be ok to change those files or do I need to submit a PR? Thanks, Richard

**Comment by jitendrapurohit** [comment]
test this please

**Comment by JoeMurray** [comment]
So I don't think @magnolia61 suggestion 2 is something that should go into this. It's a bit of a preference thing that deserves a separate discussion.

I like the idea of having a balance due field displayed somewhere close to the Pay Now button. I'm of two minds about whether it should be included. On balance, I think it makes sense to include in the same PR.

I don't think the Pay Now should be inside list of payments for a contribution since it will require opening that, and mostly people will not have partial payments, and if they do then they still don't have a need to see the list of payments.

I have not setup a test site with the PR. Could a screenshot from #3 without error be posted here? I just want to ensure that the general layout of a CiviCRM page is being respected. Might be easiest to assess with a default theme from Drupal.

**Comment by JoeMurray** [comment]
Thanks, @jitendrapurohit .

The non-error form looks fine.

Regarding error handling, the general pattern is that friendly messages should be displayed for non-programming errors, and ugly program level error feedback is acceptable for bugs. The scenario of a user attempting to pay for a partial amount on fully paid contribution is a bit in the middle: it's possible to get there through inappropriate use of back button or multiple tabs. I suppose it's acceptable to get away with an unfriendly display, though a friendly message display format would be preferred. More appropriate text would be 'This contribution is already fully paid.'

**Comment by mattwire** [comment]
> More appropriate text would be 'This contribution is already fully paid.'
@JoeMurray @jitendrapurohit I agree.  Is there any need to get rid of the statusBounce - much more user friendly, I dread when demo-ing CiviCRM and the "Sorry due to an error..." screen appears.

**Comment by jitendrapurohit** [comment]
@mattwire statusBounce is what was used before but seems it doesn't show the required error to anonymous user and is simply redirected to the drupal homepage.

@JoeMurray @magnolia61 Balance is now added to the dashboard -

image

**Comment by JoeMurray** [comment]
This is looking very good.

Just checking that you don't put a Pay Now for Pending (Incomplete), since generally that will mean the balance is in transit. 

This interface makes we wish more that we get around to refactoring the status field to better handle the fact that there may be multiple payments in process with different statuses, and a balance owing or owed as well. But that's too much to put onto this nice improvement. @eileenmcnaughton @lcdservices

**Comment by magnolia61** [comment]
Hmm. For some reason https://github.com/civicrm/civicrm-core/pull/12319.patch seems to create an incomplete patch for me. Would you be able to provide a patch file for the current PR?

**Comment by jitendrapurohit** [comment]
@magnolia61 I've squashed the changes into a single commit so the above link should work for you now -  https://github.com/civicrm/civicrm-core/pull/12319.patch

**Comment by jitendrapurohit** [comment]
test this please

**Comment by jitendrapurohit** [comment]
jenkins, retest this please

**Comment by magnolia61** [comment]
OK, last screen was caused by the Errorreporting extension. 
`ArgumentCountError: Too few arguments to function reporterror_setting_get(), 1 passed in /var/www/vhosts/xyz/webroot/sites/all/modules/civicrm_extensions/ca.bidon.reporterror/CRM/ReportError/Utils.php on line 49 and at least 2 expected in reporterror_setting_get() (line 392 of /var/www/vhosts/xyz/webroot/sites/all/modules/civicrm_extensions/ca.bidon.reporterror/reporterror.php).`

When I disabled that extension I saw the expected, but not so user friendly error screen.

screenshot from 2018-10-19 14-00-16
 
So basically, all this issue was about seems to function correctly. A nicer error message has a high priority to my persional oppinion but does not have to be part of this PR.

**Comment by JoeMurray** [comment]
I think the error text is fine but it should be a friendly error message as another has encountered it.

**Comment by magnolia61** [comment]
> I think the error text is fine but it should be a friendly error message as another has encountered it.

I agree! The new text is great, very descriptive for non-tech endusers. But the way it is presented is pretty scary. Would it be easy to present it more friendly? (and present it to anonymous users at all), or would it be better to first get the current code merged?

I think CiviCRM needs a blank placeholder error/message page where messages can be shown both to logged in, checksum and anonymous users to all get the same (error)messages.

**Comment by magnolia61** [comment]
One other observation: I'm not sure if this is a core issue or should be attributed to our payment processor (Omnipay with Mollie), but after a payment of a partial paid contribution no email is being sent.
I opened an issue for this on the Omnipay Processor side (https://github.com/eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor/issues/58) but I would appreciate is when someone confirms the email is being sent with another payment processor in this scenario.

**Comment by JoeMurray** [comment]
> One other observation: I'm not sure if this is a core issue or should be attributed to our payment processor (Omnipay with Mollie), but after a payment of a partial paid contribution no email is being sent.
> 
> I opened an issue for this on the Omnipay Processor side ([eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor#58](https://github.com/eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor/issues/58)) but I would appreciate is when someone confirms the email is being sent with another payment processor in this scenario. This would answer my following questions: "Is an confirmation mail or receipt expected to be sent after full payment of a partially paid contribution? If not, should we?"

I'm not sure what may have caused that. Could you create a separate ticket for it?

File: CRM/Contribute/BAO/Contribution.php [hunk 3]
```diff
@@ -3824,6 +3824,37 @@ public static function validateFinancialType($financialTypeId, $relationName = '
...
+   *   is the invoice contribution id (got created after processing participant payment).
+   * @param array $params
+   *   to take user provided input of transaction details.
+   * @param string $paymentType
+   *   'owed' for purpose of recording partial payments, 'refund' for purpose of recording refund payments.
+   * @param int $participantId
+   *
+   * @return array
+   */
+  public static function processAdditionalPayment($contributionId, $params, $paymentType = 'owed', $participantId = NULL) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: I was looking at this & the payment.create api - which should be basically the same function - and there doesn't seem to be much overlap!

**Comment by eileenmcnaughton** [comment]
Regarding the question of emails going out - I think if you used the Payment.create api that might trigger the emails (other conditions being met) - the code extracted here doesn't seem to be intended to do that.

File: CRM/Contribute/BAO/Contribution.php (hunk 3 above)
Comment (jitendrapurohit) [reply CRM/Contribute/BAO/Contribution.php]: Since this is the same path that we follow while doing partial payment from the back end forms, this PR also follows the same route for the front end payments too. I've looked into payment create api and wasn't able to see any param that actually triggers the correct receipt to the users.

**Comment by kcristiano** [comment]
I have tested this PR applying to both 5.7 and 5.8(master).   

From a functional standpoint it works well.  It's a feature a number of our clients desire and I would love to have a user facing form available to do this.

I understand @eileenmcnaughton 's reservations.   Is there a way we could simplify the contribution form process by requiring a form just for this process or by having a simple built in form just for this purpose?   Would that make folks more comfortable?

File: CRM/Contribute/BAO/Contribution.php (hunk 3 above)
Comment (JoeMurray) [reply CRM/Contribute/BAO/Contribution.php]: @monishdeb could you suggest to @jitendrapurohit the appropriate calls so there is good code reuse?

**Comment by eileenmcnaughton** [comment]
@kcristiano thanks for that - I am going to take a proper look when I can set aside a chunk of time

**Comment by JoeMurray** [comment]
For memberships we put in a catchall form for renewals (https://dmaster.demo.civicrm.org/civicrm/admin/setting/preferences/member?reset=1). For Pay Balance, we need to support any priceset, which complicates things a bit. 

I think this should be treated as a frontend form since it is user facing.

I am not able to comment well without further work on whether a lightweight form for this purpose would be possible and better than a normal contribution form. We want avoid if possible adding any (more) complexity to regular contribution forms. We had intended the creation of API's like order and payment to facilitate doing things other ways, like a simple Pay Balance form. I would prefer to improve the BAO calls associated with Payment API than to further complexify contribution form. Very unfortunately, this is way at the end of the development and review process rather than an early consultation on the design pattern to use. Still, I think many would like this and there might be resources to help with a refactoring/reimplementation if that is what we want.

**Comment by eileenmcnaughton** [comment]
@JoeMurray why do we need to support a price set = are we not adding a payment against an existing contribution?

**Comment by JoeMurray** [comment]
Sorry for not coming back to this discussion. 

> @JoeMurray why do we need to support a price set = are we not adding a payment against an existing contribution?

The approach I was imagining was to have a single Contribution Page that would handle all Pay Now payments. Some of the contributions might have been setup as backoffice contributions using Price sets that are not configured for any Contribution Page. In addition, they may have been edited after a partial payment, so that some of the line items have been fully paid, some partially paid, some not paid at all. I was imagining that the user needed to be shown on the Pay Now contribution page what they are paying for, which would involve using the appropriate price set. Having a contribution page able to handle any priceset, rather than just the one configured for it, would be a bit of a change in what is generally an area of toxic code.

Lots of assumptions in there. Glad to be shown I'm wrong or that there is a better way.

Sorry again for the delay.

File: CRM/Contribute/BAO/Contribution.php [hunk 4]
```diff
@@ -4519,10 +4550,38 @@ public static function isSingleLineItem($id) {
    * @param int $recur
    * @param CRM_Contribute_BAO_Contribution $contribution
    *
-   * @return array
+   * @return NULL|array
    */
   public static function completeOrder(&$input, &$ids, $objects, $transaction, $recur, $contribution) {
     $primaryContributionID = isset($contribution->id) ? $contribution->id : $objects['first_contribution']->id;
+    $paymentProcessorId = '';
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: OK - so this is my big issue here - the completeOrder function should NOT be handling partial payments. It is the function to act when a contribution has been completed and the various things need to be transitioned.

The way we should be developing here is building out the Payment.create api - which calls this completeOrder function when it is the final payment - so ideally we want the form function

completeTransaction

to call Payment.create instead of contribution.Completetransaction - which will add payments & complete the transaction if required.

Currently Payment.create currently has some gaps :-( but I think we should be addresssing those. We definitely should not be squeezing Payment.create functionality into completetransaction / completeOrder

File: CRM/Contribute/Form/Contribution/Main.php [hunk 5]
```diff
@@ -1310,20 +1310,30 @@ public function assignFormVariablesByContributionID() {
       return;
     }
     if (!$this->getContactID()) {
-      CRM_Core_Error::statusBounce(ts("Returning since there is no contact attached to this contribution id."));
+      throw new CRM_Core_Exception(ts("Returning since there is no contact attached to this contribution id."));
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/Contribution/Main.php]: I'm a bit torn here - I think it is less rude to bounce them?

File: CRM/Contribute/Form/Contribution/Main.php [hunk 6]
```diff
@@ -1310,20 +1310,30 @@ public function assignFormVariablesByContributionID() {
...
-    if (empty($payment['balance'])) {
-      CRM_Core_Error::statusBounce(ts("Returning since contribution has already been handled."));
+    if (empty($payment['balance']) || !in_array($payment['contribution_status'], array('Pending', 'Partially paid'))) {
+      throw new CRM_Core_Exception(ts("This contribution is already fully paid."));
     }
-    if (!empty($payment['total'])) {
-      $this->_pendingAmount = $payment['total'];
+
+    $paymentInfo = CRM_Core_BAO_FinancialTrxn::getPartialPaymentWithType($this->_ccid, 'contribution');
+    $pendingAmount = CRM_Utils_Array::value('amount_owed', $paymentInfo);
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/Contribution/Main.php]: I looked specifically at this part and I think we can make a preliminary change to support this - see 

https://github.com/civicrm/civicrm-core/pull/13151

which means we will be supporting the 'correct' amount at this point.

File: CRM/Contribute/Selector/Search.php [hunk 7]
```diff
@@ -412,6 +412,11 @@ public function &getRows($action, $offset, $rowCount, $sort, $output = NULL) {
       $row['contribution_status_name'] = CRM_Utils_Array::value($row['contribution_status_id'],
         $contributionStatuses
       );
+      $invoiceSettings = Civi::settings()->get('contribution_invoice_settings');
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Selector/Search.php]: This will need to change but hopefully we can separate the Search from the UserDashboard since we don't want to be doing this on all search results

File: CRM/Core/BAO/FinancialTrxn.php (hunk 1 above)
Comment (monishdeb) [reply CRM/Core/BAO/FinancialTrxn.php]: Apprently it doesn't seem odd to me consider Refunded status. But I checked the code for its impact and this function get called at two places:
```CRM//Core/BAO/FinancialTrxn.php:536:    $paid = CRM_Core_BAO_FinancialTrxn::getTotalPayments($params['contribution_id']);
CRM//Price/BAO/LineItem.php:1176:    $paidAmount = CRM_Core_BAO_FinancialTrxn::getTotalPayments($contributionId);
```
As per the usage I played with some Event's change-fee scenario  and line-item editor to check how it behaves but didn't found any unintended behaivour. So this seems fine to me. 

(Although it would be ok to get rid of the status subclause, but let's not tweak this rule in this PR and it might cause regression)

File: CRM/Contribute/Form/Contribution/Main.php (hunk 5 above)
Comment (monishdeb) [reply CRM/Contribute/Form/Contribution/Main.php]: I think previous statusBounce was correct as per the Error message which says *Returning*.. Or I would suggest:
```
CRM_Core_Error::statusBounce(ts("There is no contact attached to this contribution id."));
```

File: CRM/Contribute/BAO/Contribution.php (hunk 3 above)
Comment (monishdeb) [reply CRM/Contribute/BAO/Contribution.php]: @eileenmcnaughton @JoeMurray @jitendrapurohit Instead of having a separate function we can call CRM_Contribute_BAO_Contribution::addPayments(..) inside recordAdditionalPayment(..), ONLY for partial payment. But then addPayments(..) itself is only responsible to create proportional entries. Here's my [patch](https://gist.github.com/monishdeb/8dc327fb3f7765729d554630cfc72528) to does this change. You can then call this same function for online form too.

On the other hand, in order to extend Payment.create API to handle partial payment and even refund, we must have payment details in the ```$params``` and eventually it will call processAdditionalPayment(...) to handle additional owed or refund amount, the intent is to move the code towards BAO fn from API. What do you think?

File: CRM/Contribute/BAO/Contribution.php (hunk 4 above)
Comment (monishdeb) [reply CRM/Contribute/BAO/Contribution.php]: As mentioned above once, we can make the existing fn ```CRM_Contribute_BAO_Contribution::recordAdditonalPayment(..)``` more robust to handle partial and even refund payment properly then we can call this fn in Payment.Create API, in this way we can differentiate it from completetransaction

File: CRM/Contribute/Selector/Search.php (hunk 7 above)
Comment (monishdeb) [reply CRM/Contribute/Selector/Search.php]: There is a big obstruction in implementing this separation as both the Search and UserDashboard uses the same Selector to build search query and assign its result to template. Unless if we can decide the returnproperties / select elements for respective forms, something like
```patch
$controller = new CRM_Core_Controller_Simple(
      'CRM_Contribute_Form_Search',
      ts('Contributions'),
      NULL,
      FALSE, FALSE, TRUE, FALSE
    );
    $controller->setEmbedded(TRUE);
    $controller->reset();
    $controller->set('limit', 12);
    $controller->set('cid', $this->_contactId);
 +   $controller->set('returnproperties', [
 +       'contact_id',
 +   'contribution_id',
 +   'contact_type',
 +   'sort_name',
 +   'amount_level',
 +   'total_amount',
 +   'financial_type',
  +   'balance',
  +    ]);
    $controller->set('context', 'user');
    $controller->set('force', 1);
    $controller->process();
    $controller->run();
```
At [here](https://github.com/civicrm/civicrm-core/blob/master/CRM/Contribute/Page/UserDashboard.php#L49)

Comment (eileenmcnaughton) [reply CRM/Contribute/Selector/Search.php]: @monishdeb there is another PR here https://github.com/civicrm/civicrm-core/pull/13147 just to stop using the search controller for the dashboard - my primary reason was performance & reducing use of a complex & poorly understood code construct but it also helps here

Note the search controller is something that we should aim to deprecate from everywhere except direct searches IMHO & then potentially either replace there or leap- by -extension away from it

File: CRM/Contribute/BAO/Contribution.php (hunk 3 above)
Comment (eileenmcnaughton) [reply CRM/Contribute/BAO/Contribution.php]: @monishdeb right - it seems Payment.create api needs to accept more parameters. The goal is the code logic should not be in that function - but that should be the function we call instead of a BAO function from any form.

File: CRM/Contribute/BAO/Contribution.php (hunk 4 above)
Comment (jitendrapurohit) [reply CRM/Contribute/BAO/Contribution.php]: Hmm, concur with the improvement of the payment create api to include more flexibility in accepting the params 👍  

As per the changes in this PR - looking at the screenshot in the PR description and some following comments - the front end form available to the user displays a non-editable balance amount which on submission do completes the payment for the contribution, hence the logic was kept same as we did the implementation for completing `pending` payments.

But, as the backend form offers a set of periodic payments until a contribution is completed (eg $100 is completed in $20, $30 and then $50) - agree this should also be added for front end forms too - which was also asked by our client after this PR and we implemented for them but thought to create a new PR for that and not mess this up as it had already completed some testing by Richard and others.

I can add the improvements related to create payment api, etc by the end of this week 👍 

Thanks Eileen and Monish for the comments.

Comment (eileenmcnaughton) [reply CRM/Contribute/BAO/Contribution.php]: So my feeling is we should focus on getting our basics right first. Ie. we want to get the payment api working well enough that we can replace all instances where a payment is received with Payment.create (this would include the many places where we currently call completetransaction).

Fortunately we have built up pretty substantial test 'assets' in most of the places that receive payments.

I would like to propose that we work on maturing the Payment.create api by switching the AdditionalPayment form over to using it as a first step. I did a [commit  here](https://github.com/eileenmcnaughton/civicrm-core/commit/431a5909dc800459d2de3a9626d3b0d0a51e340b) - the current failure reason is some sillyness in the Payment create api itself (it's not updating status when you update a partially paid) but I think we can add a test directly against the Payment.create api that captures that scenario & fixes it.



I think really cleaning up Payment.create & working together on switching forms to it will set us up a lot better in the future  - both while we are working with quick form & for when we finally escape it.

Comment (jitendrapurohit) [reply CRM/Contribute/BAO/Contribution.php]: Updated the PR to use the payment create API for partial payments on front and backend forms. I've also included the above commit from @eileenmcnaughton. Till now, I don't see any issues, but changing this to wip until the following point is completely checked.

- Test payment create api for any gaps and handle them. Eg. it does not send email receipt from front-end forms after the thankyou page is loaded.
- Handle the assignment of `Civi::settings()->get('contribution_invoice_settings');`
- Add more unit test to assert the API results.

File: CRM/Contribute/Form/Contribution/Main.php (hunk 6 above)
Comment (jitendrapurohit) [reply CRM/Contribute/Form/Contribution/Main.php]: Cool - as @monishdeb is already reviewing that - will update the call here as soon as it is merged.

File: CRM/Contribute/Form/Contribution/Main.php (hunk 5 above)
Comment (jitendrapurohit) [reply CRM/Contribute/Form/Contribution/Main.php]: Hmm, statusBounce is what was used before. This was modified to satisfy @magnolia61 requirement [in this comment](https://github.com/civicrm/civicrm-core/pull/12319#issuecomment-398029244) of not seeing any error when the user is navigated to CMS home page after the bounce is triggered. I'm

Comment (magnolia61) [reply CRM/Contribute/Form/Contribution/Main.php]: It was because a message to users clicking a Pay Now checksum link which was already paid or expired got no message at all. With all improvement ideas, I think it could be a separate issue to provide a end-user friendly error page for non-logged users.

**Comment by JoeMurray** [comment]
Many thanks for this work, @jitendrapurohit 

@monishdeb I know you're very busy these days, but could you schedule a review of this for Wednesday or Thursday? Thanks.

**Comment by magnolia61** [comment]
I tested out the latest PR with the payment api functionality.
I found several bugs worth mentioning.

scenario 1:
contribution 50 euro. with (frontend) Pay Now checksum link I pay 30. 
result: the total sum of the contribution changes from 50 to 30. 
desired result: contribution total should remain 50, but a payment of 30 should be recorded

scenario 2:
event price set
a) basic fee (50 euro)
b) optional fee (30 euro)

Payment is complete for basic fee. Event price set is changed (backend) to include the optional fee. There is now 30 euro (the optional fee) still to be paid.
I pay with the checksum Pay Now link. 
result: an extra contribution of 30 euro is recorded. 
desired result: a payment towards the original contribution is expected.

I will test some more scenarios but i thought I'd let you know this already

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit can we break any refactorings into small PRs that only do the refactorings & merge those first to get the amount of stuff happening in here down. I'd rather chip away at this a bit  - e.g try out getting something like this down https://github.com/eileenmcnaughton/civicrm-core/commit/431a5909dc800459d2de3a9626d3b0d0a51e340b first with changes in payment.create first & mature the payment.create api as step one

**Comment by jitendrapurohit** [comment]
Sorry - back from a small break and looking into this comment. @magnolia61 Seems like the usecase replicates for "pending" payments only and works fine for partial payment I tested in my above comment. I've pushed a minor update to this PR and it fixes your cases too.

@eileenmcnaughton Sure. I'll create some PR in batches and lets merge them first.

**Comment by magnolia61** [comment]
Hello @jitendrapurohit just tested your updated PR and it indeed prevents a new contribution to be created. Unfortunately the Pay Now screen now shows the full amount instead of the Balance

INITIAL ACTION:	Registered for event, basic fee of 50 euro selected and paid
INITIAL RESULT:	Contribution recorded, status: Completed
BACKEND ACTION: Backend registration edited to include the options.
BACKEND RESULT:Contribution Status changes to: Partially Paid. Total Fee: 80 euro. Balance: 30 euro
USER ACTION:	User uses Pay Now Checksum link to pay for remaining 30 euro
ACTUAL RESULT:	User sees payment screen for the full 80 euro (not just the balance)

screenshot from 2018-12-17 12-18-29
screenshot from 2018-12-17 12-17-50

**Comment by jitendrapurohit** [comment]
@magnolia61 This does not seem to be true on my testing. See the below screenshot where a contribution is created for $50(completed) -> selections changed to include $30 (status changes to partially paid) -> Navigating to the pay now link gives the below screen. Do you see any points missing here?

image

Also scenario 1 works fine on my end.

Side Note: I recall you had problems in fetching the correct patch before, so just need to confirm if you have the correct updated patch applied on your review setup, i.e, https://patch-diff.githubusercontent.com/raw/civicrm/civicrm-core/pull/12319.diff?

**Comment by magnolia61** [comment]
Maybe @jitendrapurohit I will double-doublecheck. What I did was a 
```
wget https://github.com/civicrm/civicrm-core/pull/12319.patch
patch -p1 < 12319.patch
```
Will let you know my results in a few moments

**Comment by monishdeb** [comment]
@magnolia61 sometimes if there's a patch file already present with the same name then it creates 12319.patch.1 without raising a warning. Please ensure that there is no 12319.patch present in 
Civi dir before doing ```wget```

**Comment by jitendrapurohit** [comment]
I've raised a PR https://github.com/civicrm/civicrm-core/pull/13330 which only contains fix for the backend forms and uses payment create API for the payment recorded via admin interface. It moves the code from API -> BAO level and also focuses on making the `emailReceipt()` as a common function which can also be accessed by front end contribution pages to send receipts.

**Comment by magnolia61** [comment]
Hello Jitendra, I am aware of your new backend only PR (I have tested it and commented) and also Eileens recent work. 
Still I have some feedback for testing this PR which is mostly working ok for us, and I think my experience might help development.

When I use this PR 12319 on civicrm 5.9 or master 'normal' contributions work fine. It is the partial contributions that when correctly processed by the payment processor returns to the payment screen with the message: "Your payment was not successful. Please try again".

screenshot from 2018-12-31 16-02-28

**Comment by magnolia61** [comment]
I am very happy to see all that is going on with the payment.create function which has it's origin in this mother-issue on front end partial payments. Hat I am missing is a gltlab issue to track all the sub-PRs for a bit of overview. I am collecting all relevant PR's in this place. @eileenmcnaughton : is there already a gltlab issue for this that I oversaw?

**Original issue:**
https://github.com/civicrm/civicrm-core/pull/12319 Extend pay now functionality for partially paid contribution

**Steps to first create a proper Payment api**
https://github.com/civicrm/civicrm-core/pull/13330 [WIP] Modify backend Record Payment forms to use payment create API 
https://github.com/civicrm/civicrm-core/pull/13370 extract internals of Payment.create into function on BAO class
https://github.com/civicrm/civicrm-core/pull/13647 Modify backend Record Payment forms to use payment create API 
https://github.com/civicrm/civicrm-core/pull/13689 Call Payment.create from payment.cancel
https://github.com/civicrm/civicrm-core/pull/13690 Clean up Payment.create function

**Issues on payment sendconfirmation:**
https://github.com/civicrm/civicrm-core/pull/13561 Add new Payment.sendconfirmation api 
https://github.com/civicrm/civicrm-core/pull/13609 Payment.sendconfirmation api - add further tpl variables
https://github.com/civicrm/civicrm-core/pull/13610 Payment.sendconfirmation api - add further tpl variables
https://github.com/civicrm/civicrm-core/pull/13647 Modify backend Record Payment forms to use payment create API (not yet merged)
https://github.com/civicrm/civicrm-core/pull/13649 Switch additional payment form to use Payment.sendconfirmation api (merged)

**Comment by eileenmcnaughton** [comment]
@magnolia61 thanks for bringing them all together! Yes, we are just trying to clean up the code first so we can do this properly - there is a long legacy of difficult code here :-( Over probably the past 2 years I think Monish, Pradeep & I have built up a pretty good amount of tests on the code which means we are in a position to tidy it up now - we're not quite there with the payment.create fixes yet though - a few more rounds left

**Comment by eileenmcnaughton** [comment]
I'm going to close this as it is stale & has become more of an issue tracker for 'cleanup payment code to the point where we can fix this and related things reliably'

Currently the open work on this is this PR https://github.com/civicrm/civicrm-core/pull/13689

& once that is merged I'll look at finishing off the next parts of the payment create consolidation

**Comment by magnolia61** [comment]
@jitendrapurohit Would it be possible to rebase this patch to work with 5.14? Or would the whole frontend partial payment patch need to be rewritten using all latest payment create api functionality?

**Comment by eileenmcnaughton** [comment]
@magnolia61 the changes needed will have changed quite a bit by now.  This patch here is the next one unmerged to take us forward in this process https://github.com/civicrm/civicrm-core/pull/14408

Do we need to load a contribution page for pay now or could we just load the add payment form (as in the backend) which would be easier & less risky

**Comment by magnolia61** [comment]
> Do we need to load a contribution page for pay now or could we just load the add payment form (as in the backend) which would be easier & less risky

I think the lean and mean add payment form would be very much ok for this. 
There are different possibilities to use it:
a) The enduser could be able to enter an amount to pay 
b) The amount could be locked readonly to be the full amount
c) The full amount would be entered, but the user would be able to change it 

I think I would prefer option c)
Probably changing from a full contribution page to the add payment form makes all of this much less complicated :-)

**Comment by magnolia61** [comment]
I have opened up a new pr that enables the Pay Now button and payment via checksum link for partially paid contributions. Still a work in progress as there are a few minor issues. But the main nice preliminary conclusion is that all of the hard work Eileen put in crafting the payment create api makes it kinda work out of the box. Please take a look at https://github.com/civicrm/civicrm-core/pull/15696 and if possible help me move this forward as it would be great to be able to include it in 5.20 which includes a lot of financial cleanup and improvements ;-)

**Comment by magnolia61** [comment]
I opened an issue to advocate for an improvement to proceed with using Record Payment form for the Pay Now functionality
https://lab.civicrm.org/dev/core/-/issues/2095 I cannot code this myself (no skills) but hope someone working on related stuff can drop by and make the Pay Now use this function properly. I am available to help test this stuff anytime :-)
